ext/libp2p/peer.rs

```diff
@@ -422,56 +468,62 @@ mod tests {
         init();
         let cancellation_token = CancellationToken::new();
 
-        let server_id_keys = identity::Keypair::generate_ed25519();
-        let server_peer_id = server_id_keys.public().to_peer_id();
-        let mut server_transport = create_transport(&server_id_keys).unwrap();
+        let listener_id_keys = identity::Keypair::generate_ed25519();
+        let listener_peer_id = listener_id_keys.public().to_peer_id();
+        let listener_transport = create_transport(&listener_id_keys).unwrap();
+
+        let listener_behavior = {
+            #[derive(NetworkBehaviour)]
+            struct ListenerBehaviour {
+                pub ping: libp2p::ping::Behaviour,
+                pub keep_alive: libp2p::swarm::keep_alive::Behaviour,
+            }
+            ListenerBehaviour {
+                ping: libp2p::ping::Behaviour::new(libp2p::ping::Config::new()),
+                keep_alive: libp2p::swarm::keep_alive::Behaviour,
+            }
+        };
+
+        let mut listener_swarm =
+            Swarm::with_tokio_executor(listener_transport, listener_behavior, listener_peer_id);
 
         // FIXME: Use an ephemeral port number here.
         // Listen on port 0, read back the port assigned by the OS
-        let server_addr: Multiaddr = "/ip4/127.0.0.1/tcp/10458".parse().unwrap();
-        server_transport.listen_on(server_addr.clone()).unwrap();
-
-        let mut server_swarm = Swarm::with_tokio_executor(
-            server_transport,
-            libp2p::ping::Behaviour::new(
-                libp2p::ping::Config::new()
-                    .with_max_failures(std::num::NonZeroU32::new(10).unwrap()),
-            ),
-            server_peer_id,
-        );
-        let server_task = {
+        let listener_addr: Multiaddr = "/ip4/127.0.0.1/tcp/10458".parse().unwrap();
+        listener_swarm.listen_on(listener_addr.clone()).unwrap();
```

Comment:
@mxinden What is the canonical way for setting up a Swarm listening on an ephemeral port?

I want to write something like this - notice the port is 0 (`/tcp/0`):

```rs
let listener_addr: Multiaddr = "/ip4/127.0.0.1/tcp/10458".parse().unwrap();
listener_swarm.listen_on(listener_addr.clone()).unwrap();
// 1. wait until OS binds the port a specific number
// 2. get the updated multi address
// E.g. the function implementing (1) can return a future of the new multiaddr
```

Reply:
You can call `listen_on("/ip4/127.0.0.1/tcp/0".parse().unwrap()).unwrap()` and then wait for a `SwarmEvent::NewListenAddr`.